{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jrqPkptsW5Vs"
      },
      "source": [
        "# **scNET: Learning Context-Specific Gene and Cell Embeddings by Integrating Single-Cell Gene Expression Data with Protein-Protein Interaction Information**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RTNzcaI1CiWZ"
      },
      "source": [
        "**Install scnet using pip**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-v0IduW97F8f"
      },
      "source": [
        "**It may be necessary to restart (not delete) the runtime after installing scNET in Google Colab. We are working on fixing this issue in future releases.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "jl04t1P9pBYt",
        "outputId": "54394037-90dc-49ac-e715-5d60a494da05",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: scnet in /usr/local/lib/python3.12/dist-packages (0.2.2.6)\n",
            "Requirement already satisfied: torch==2.6.0 in /usr/local/lib/python3.12/dist-packages (from scnet) (2.6.0)\n",
            "Requirement already satisfied: torch-geometric==2.6.1 in /usr/local/lib/python3.12/dist-packages (from scnet) (2.6.1)\n",
            "Requirement already satisfied: pandas>=2.2.1 in /usr/local/lib/python3.12/dist-packages (from scnet) (2.2.2)\n",
            "Requirement already satisfied: numpy==1.26.4 in /usr/local/lib/python3.12/dist-packages (from scnet) (1.26.4)\n",
            "Requirement already satisfied: networkx>=3.1 in /usr/local/lib/python3.12/dist-packages (from scnet) (3.5)\n",
            "Requirement already satisfied: scanpy>=1.11.0 in /usr/local/lib/python3.12/dist-packages (from scnet) (1.11.5)\n",
            "Requirement already satisfied: scikit-learn>=1.4.1 in /usr/local/lib/python3.12/dist-packages (from scnet) (1.6.1)\n",
            "Requirement already satisfied: gseapy>=1.1.6 in /usr/local/lib/python3.12/dist-packages (from scnet) (1.1.10)\n",
            "Requirement already satisfied: matplotlib>=3.8.0 in /usr/local/lib/python3.12/dist-packages (from scnet) (3.10.0)\n",
            "Requirement already satisfied: igraph in /usr/local/lib/python3.12/dist-packages (from scnet) (1.0.0)\n",
            "Requirement already satisfied: leidenalg in /usr/local/lib/python3.12/dist-packages (from scnet) (0.11.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.12/dist-packages (from scnet) (4.67.1)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (3.20.0)\n",
            "Requirement already satisfied: typing-extensions>=4.10.0 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (4.15.0)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (3.1.6)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (2025.3.0)\n",
            "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.4.127 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.4.127 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.4.127 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cudnn-cu12==9.1.0.70 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (9.1.0.70)\n",
            "Requirement already satisfied: nvidia-cublas-cu12==12.4.5.8 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (12.4.5.8)\n",
            "Requirement already satisfied: nvidia-cufft-cu12==11.2.1.3 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (11.2.1.3)\n",
            "Requirement already satisfied: nvidia-curand-cu12==10.3.5.147 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (10.3.5.147)\n",
            "Requirement already satisfied: nvidia-cusolver-cu12==11.6.1.9 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (11.6.1.9)\n",
            "Requirement already satisfied: nvidia-cusparse-cu12==12.3.1.170 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (12.3.1.170)\n",
            "Requirement already satisfied: nvidia-cusparselt-cu12==0.6.2 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (0.6.2)\n",
            "Requirement already satisfied: nvidia-nccl-cu12==2.21.5 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (2.21.5)\n",
            "Requirement already satisfied: nvidia-nvtx-cu12==12.4.127 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (12.4.127)\n",
            "Requirement already satisfied: nvidia-nvjitlink-cu12==12.4.127 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (12.4.127)\n",
            "Requirement already satisfied: triton==3.2.0 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (3.2.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (75.2.0)\n",
            "Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.12/dist-packages (from torch==2.6.0->scnet) (1.13.1)\n",
            "Requirement already satisfied: aiohttp in /usr/local/lib/python3.12/dist-packages (from torch-geometric==2.6.1->scnet) (3.13.2)\n",
            "Requirement already satisfied: psutil>=5.8.0 in /usr/local/lib/python3.12/dist-packages (from torch-geometric==2.6.1->scnet) (5.9.5)\n",
            "Requirement already satisfied: pyparsing in /usr/local/lib/python3.12/dist-packages (from torch-geometric==2.6.1->scnet) (3.2.5)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.12/dist-packages (from torch-geometric==2.6.1->scnet) (2.32.4)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.12/dist-packages (from sympy==1.13.1->torch==2.6.0->scnet) (1.3.0)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.12/dist-packages (from gseapy>=1.1.6->scnet) (1.16.3)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.8.0->scnet) (1.3.3)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.8.0->scnet) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.8.0->scnet) (4.60.1)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.8.0->scnet) (1.4.9)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.8.0->scnet) (25.0)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.8.0->scnet) (11.3.0)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.12/dist-packages (from matplotlib>=3.8.0->scnet) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.12/dist-packages (from pandas>=2.2.1->scnet) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.12/dist-packages (from pandas>=2.2.1->scnet) (2025.2)\n",
            "Requirement already satisfied: anndata>=0.8 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (0.12.6)\n",
            "Requirement already satisfied: h5py>=3.7.0 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (3.15.1)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (1.5.2)\n",
            "Requirement already satisfied: legacy-api-wrap>=1.4.1 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (1.5)\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (8.4.0)\n",
            "Requirement already satisfied: numba!=0.62.0rc1,>=0.57.1 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (0.60.0)\n",
            "Requirement already satisfied: patsy!=1.0.0 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (1.0.2)\n",
            "Requirement already satisfied: pynndescent>=0.5.13 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (0.5.13)\n",
            "Requirement already satisfied: seaborn>=0.13.2 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (0.13.2)\n",
            "Requirement already satisfied: session-info2 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (0.2.3)\n",
            "Requirement already satisfied: statsmodels>=0.14.5 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (0.14.5)\n",
            "Requirement already satisfied: umap-learn>=0.5.6 in /usr/local/lib/python3.12/dist-packages (from scanpy>=1.11.0->scnet) (0.5.9.post2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.12/dist-packages (from scikit-learn>=1.4.1->scnet) (3.6.0)\n",
            "Requirement already satisfied: texttable>=1.6.2 in /usr/local/lib/python3.12/dist-packages (from igraph->scnet) (1.7.0)\n",
            "Requirement already satisfied: array-api-compat>=1.7.1 in /usr/local/lib/python3.12/dist-packages (from anndata>=0.8->scanpy>=1.11.0->scnet) (1.12.0)\n",
            "Requirement already satisfied: zarr!=3.0.*,>=2.18.7 in /usr/local/lib/python3.12/dist-packages (from anndata>=0.8->scanpy>=1.11.0->scnet) (3.1.3)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.12/dist-packages (from numba!=0.62.0rc1,>=0.57.1->scanpy>=1.11.0->scnet) (0.43.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.12/dist-packages (from python-dateutil>=2.7->matplotlib>=3.8.0->scnet) (1.17.0)\n",
            "Requirement already satisfied: aiohappyeyeballs>=2.5.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp->torch-geometric==2.6.1->scnet) (2.6.1)\n",
            "Requirement already satisfied: aiosignal>=1.4.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp->torch-geometric==2.6.1->scnet) (1.4.0)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp->torch-geometric==2.6.1->scnet) (25.4.0)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.12/dist-packages (from aiohttp->torch-geometric==2.6.1->scnet) (1.8.0)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.12/dist-packages (from aiohttp->torch-geometric==2.6.1->scnet) (6.7.0)\n",
            "Requirement already satisfied: propcache>=0.2.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp->torch-geometric==2.6.1->scnet) (0.4.1)\n",
            "Requirement already satisfied: yarl<2.0,>=1.17.0 in /usr/local/lib/python3.12/dist-packages (from aiohttp->torch-geometric==2.6.1->scnet) (1.22.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.12/dist-packages (from jinja2->torch==2.6.0->scnet) (3.0.3)\n",
            "Requirement already satisfied: charset_normalizer<4,>=2 in /usr/local/lib/python3.12/dist-packages (from requests->torch-geometric==2.6.1->scnet) (3.4.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.12/dist-packages (from requests->torch-geometric==2.6.1->scnet) (3.11)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.12/dist-packages (from requests->torch-geometric==2.6.1->scnet) (2.5.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.12/dist-packages (from requests->torch-geometric==2.6.1->scnet) (2025.10.5)\n",
            "Requirement already satisfied: donfig>=0.8 in /usr/local/lib/python3.12/dist-packages (from zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy>=1.11.0->scnet) (0.8.1.post1)\n",
            "Requirement already satisfied: numcodecs>=0.14 in /usr/local/lib/python3.12/dist-packages (from numcodecs[crc32c]>=0.14->zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy>=1.11.0->scnet) (0.16.3)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.12/dist-packages (from donfig>=0.8->zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy>=1.11.0->scnet) (6.0.3)\n",
            "Requirement already satisfied: crc32c>=2.7 in /usr/local/lib/python3.12/dist-packages (from numcodecs[crc32c]>=0.14->zarr!=3.0.*,>=2.18.7->anndata>=0.8->scanpy>=1.11.0->scnet) (2.8)\n"
          ]
        }
      ],
      "source": [
        "!pip install scnet"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hmR9bgrWB3NJ"
      },
      "source": [
        "**Download example data**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 125
        },
        "id": "O6IK04C4Icll",
        "outputId": "cb3974ea-0bee-4e82-e2f5-e02ac0bc2ca4"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Downloading...\n",
            "From (original): https://drive.google.com/uc?id=1C_G14cWk95FaDXuXoRw-caY29DlR9CPi\n",
            "From (redirected): https://drive.google.com/uc?id=1C_G14cWk95FaDXuXoRw-caY29DlR9CPi&confirm=t&uuid=59079e42-448f-4ae6-ae9a-e5ff4fab48a2\n",
            "To: /content/example.h5ad\n",
            "100%|██████████| 1.36G/1.36G [00:13<00:00, 100MB/s]\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'./example.h5ad'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 11
        }
      ],
      "source": [
        "import gdown\n",
        "download_url = f'https://drive.google.com/uc?id=1C_G14cWk95FaDXuXoRw-caY29DlR9CPi'\n",
        "output_path = './example.h5ad'\n",
        "gdown.download(download_url, output_path, quiet=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EXcsrHV8sqFO"
      },
      "source": [
        "# Import scNET and model traning"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "y3SUWWD0sqFO"
      },
      "outputs": [],
      "source": [
        "import scNET\n",
        "import scanpy as sc\n",
        "\n",
        "#For faster processing in medium to large datasets (e.g. 30K or above cells), the maximum cells batch size can be increased depending on the available GPU memory.\n",
        "\n",
        "#For GPU with 24GB memory\n",
        "#scNET.main.MAX_CELLS_BATCH_SIZE = 3000\n",
        "\n",
        "#for GPU with 40GB memory\n",
        "#scNET.main.MAX_CELLS_BATCH_SIZE = 4000\n",
        "\n",
        "#For GPU with 80GB memory or more\n",
        "#scNET.main.MAX_CELLS_BATCH_SIZE = 8000\n",
        "\n",
        "#otherwize, do not change the default value"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "2bc4oqdJsqFP"
      },
      "outputs": [],
      "source": [
        "#To control the cutoff of gene expression, the minimum precetage of cells expressing a gene can be adjusted. The default all expressed genes are considered.\n",
        "#For example, to consider genes expressed in at least 5% of cells\n",
        "#scNET.main.EXPRESSION_CUTOFF = 0.05"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "c_6938MCsqFP"
      },
      "outputs": [],
      "source": [
        "#For larger dataset (10K or above), containing larger number of subcommunities, the number of encoder layers could be increased to 4 or more. the default value is 3.\n",
        "scNET.main.NUM_LAYERS = 3"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "TsOeFPAJuo2U"
      },
      "outputs": [],
      "source": [
        "#To control the number of deifferentially expressed genes, the default value is 2000\n",
        "#For example, to consider 3500 DE genes\n",
        "scNET.main.DE_GENES_NUM = 3500\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3N7VMJjWCJvh"
      },
      "source": [
        "## run_scNET arguments:\n",
        " **obj (AnnData, optional):** AnnData obj.\n",
        "\n",
        "  **pre_processing_flag (bool, optional):** If True, perform pre-processing steps.\n",
        "\n",
        "  **human_flag (bool, optional):** Controls gene name casing in the network.\n",
        "\n",
        "  **number_of_batches (int, optional):** Number of mini-batches for the training.\n",
        "\n",
        "  **split_cells (bool, optional):** If True, split by cells instead of edges during training. If using A100 could be True (for 10K cells or less)\n",
        "\n",
        "  **n_neighbors (int, optional):** Number of neighbors for building the adjacency graph.\n",
        "\n",
        "  **max_epoch (int, optional):** Max number of epochs for model training (recommended 250-300, and for larger dataset consider also 300-350).\n",
        "\n",
        "  **model_name (str, optional):** Identifier for saving the model outputs.\n",
        "  \n",
        "  **save_model_flag (bool, optional):** If True, save the trained model."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cqHRFrT90HVX",
        "outputId": "e5c5ad7c-5fbe-4505-cd6f-ef77395b492d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "WARNING: adata.X seems to be already log-transformed.\n",
            "/usr/local/lib/python3.12/dist-packages/scNET/Data/format_h_sapiens.csv\n",
            "N genes: (7107, 9019)\n",
            "Highly variable genes: 3500\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Training: 100%|██████████| 300/300 [2:11:28<00:00, 26.30s/it]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Best Network AUC: 0.8867336125781338\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        }
      ],
      "source": [
        "obj = sc.read_h5ad(\"./example.h5ad\")\n",
        "scNET.run_scNET(obj, pre_processing_flag=False, human_flag=False, number_of_batches=10, split_cells=True, max_epoch=300, model_name = \"test\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gu_vlv2vL4bt"
      },
      "source": [
        "\n",
        "# Using the model's outputs"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "x5eVQhrbDNjN"
      },
      "source": [
        "**Load all the relevant embeddings**\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {
        "id": "QoYWucAi8Z7b"
      },
      "outputs": [],
      "source": [
        "embedded_genes, embedded_cells, node_features , out_features =  scNET.load_embeddings(\"test\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GeUbbjZZLerk"
      },
      "source": [
        "**Create Scanpy object based on the reconstructed gene expression**\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "id": "TDm7W-O5Miuo",
        "outputId": "182212c2-8192-4c8d-8413-6ca1c72ada02",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 382
        }
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "partially initialized module 'torchvision' has no attribute 'extension' (most likely due to a circular import)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-3203598809.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcell_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"0\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"Macrophages\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"1\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"Macrophages\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"2\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"CD8 Tcells\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"3\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"Microglia\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"4\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"Cancer\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"5\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"CD4 Tcells\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"6\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"B Cells\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"10\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"Prolifrating Tcells\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"8\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"Cancer\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"11\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\"NK\"\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Cell Type\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseurat_clusters\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcell_types\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mrecon_obj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscNET\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate_reconstructed_obj\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode_features\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_features\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/scNET/coEmbeddedNetwork.py\u001b[0m in \u001b[0;36mcreate_reconstructed_obj\u001b[0;34m(node_features, out_features, orignal_obj)\u001b[0m\n\u001b[1;32m     81\u001b[0m   \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mneighbors\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_neighbors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_pcs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m15\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     82\u001b[0m   \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mleiden\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mresolution\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 83\u001b[0;31m   \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mumap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     84\u001b[0m   \u001b[0;32mreturn\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     85\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/scanpy/tools/_umap.py\u001b[0m in \u001b[0;36mumap\u001b[0;34m(adata, min_dist, spread, n_components, maxiter, alpha, gamma, negative_sample_rate, init_pos, random_state, a, b, method, key_added, neighbors_key, copy)\u001b[0m\n\u001b[1;32m    179\u001b[0m         \u001b[0;31m# umap 0.5.0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    180\u001b[0m         \u001b[0mwarnings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfilterwarnings\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmessage\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34mr\"Tensorflow not installed\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 181\u001b[0;31m         \u001b[0;32mimport\u001b[0m \u001b[0mumap\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    182\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    183\u001b[0m     \u001b[0;32mfrom\u001b[0m \u001b[0mumap\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mumap_\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mfind_ab_params\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msimplicial_set_embedding\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/umap/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mcatch_warnings\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0msimplefilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m         \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mparametric_umap\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mParametricUMAP\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mload_ParametricUMAP\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     warn(\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/umap/parametric_umap.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[0;32mimport\u001b[0m \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0monnx\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 41\u001b[0;31m     \u001b[0;32mimport\u001b[0m \u001b[0mtorchvision\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     42\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0mwarn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\"\"Torch and ONNX required for exporting to those formats.\"\"\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/torchvision/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m# .extensions) before entering _meta_registrations.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mextension\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m_HAS_OPS\u001b[0m  \u001b[0;31m# usort:skip\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mtorchvision\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m_meta_registrations\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdatasets\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mops\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtransforms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mutils\u001b[0m  \u001b[0;31m# usort:skip\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/torchvision/_meta_registrations.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0;34m@\u001b[0m\u001b[0mregister_meta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"roi_align\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmeta_roi_align\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrois\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspatial_scale\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpooled_height\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpooled_width\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msampling_ratio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maligned\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_check\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrois\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mlambda\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"rois must have shape as Tensor[K, 5]\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/torchvision/_meta_registrations.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(fn)\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mregister_meta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mop_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moverload_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"default\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mtorchvision\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextension\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_has_ops\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m             \u001b[0mget_meta_lib\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimpl\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtorch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mops\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtorchvision\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mop_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moverload_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: partially initialized module 'torchvision' has no attribute 'extension' (most likely due to a circular import)"
          ]
        }
      ],
      "source": [
        "cell_types = {\"0\":\"Macrophages\",\"1\":\"Macrophages\",\"2\":\"CD8 Tcells\",\"3\":\"Microglia\",\"4\":\"Cancer\",\"5\":\"CD4 Tcells\",\"6\":\"B Cells\",\"10\":\"Prolifrating Tcells\",\"8\":\"Cancer\",\"11\":\"NK\"}\n",
        "obj.obs[\"Cell Type\"] = obj.obs.seurat_clusters.map(cell_types)\n",
        "recon_obj = scNET.create_reconstructed_obj(node_features, out_features, obj)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Rh9XrvVyLwlt"
      },
      "source": [
        "**Plot marker genes**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "aBp5-eyHHgi7"
      },
      "outputs": [],
      "source": [
        "sc.pl.umap(recon_obj, color=[\"Cell Type\",\"Cd4\",\"Cd8a\",\"Cd14\",\"Icos\",\"P2ry12\",\"Mki67\",\"Ncr1\"], show=True, legend_loc='on data')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "W9sWEFleHAxR"
      },
      "source": [
        "**Example of Calculating Marker Gene AUPR**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "OXqfU5NJvFgk"
      },
      "outputs": [],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "from sklearn.metrics import precision_recall_curve, auc\n",
        "from itertools import cycle\n",
        "\n",
        "def calculate_marker_gene_aupr(adata, marker_genes, cell_types):\n",
        "    colors = cycle(['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2'])\n",
        "    plt.figure(figsize=(10, 8))\n",
        "\n",
        "    for marker_gene, cell_type, color in zip(marker_genes, cell_types, colors):\n",
        "        gene_expression = adata[:, marker_gene].X.toarray().flatten()\n",
        "        binary_labels = (adata.obs[\"Cell Type\"].isin(cell_type)).astype(int)\n",
        "\n",
        "        precision, recall, _ = precision_recall_curve(binary_labels, gene_expression)\n",
        "        aupr = auc(recall, precision)\n",
        "        plt.plot(recall, precision, color=color, lw=2,\n",
        "                 label=f'PRAUC={aupr:.2f} for {marker_gene} ({cell_type[0]})')\n",
        "\n",
        "    plt.xlabel('Recall', fontsize=14)\n",
        "    plt.ylabel('Precision', fontsize=14)\n",
        "    plt.title('Precision-Recall Curve by Cell Type', fontsize=16)\n",
        "    plt.legend(loc=\"best\", fontsize=12)\n",
        "    plt.grid(True)\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "calculate_marker_gene_aupr(recon_obj,['Cd8a','Cd4','Cd14',\"P2ry12\",\"Ncr1\",\"Mki67\",\"Tert\"],[[\"CD8 Tcells\"],['CD4 Tcells'], ['Macrophages'], ['Microglia'], [\"NK\"],[\"Prolifrating Tcells\"],[\"Cancer\"]])\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2xKxBxmaVY20"
      },
      "source": [
        "## GO Semantic Similarity"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MTMbPaBpV5cE"
      },
      "source": [
        "Download the Semantic Similarity table calculated using the GoSemSim package in R.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "FiQ4X_qQVDC2"
      },
      "outputs": [],
      "source": [
        "import gdown\n",
        "download_url = f'https://drive.google.com/uc?id=1glnzD-rFk_zzBVF7tDmjTP-HmhqKEPVU'\n",
        "output_path = './sim.csv'\n",
        "gdown.download(download_url, output_path, quiet=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_K3xuvEbWI60"
      },
      "source": [
        "Filter out uninformative genes"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "spn6_lDuVGYF"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "sim = pd.read_csv(\"./sim.csv\", index_col=0)\n",
        "fil = (sim == 1).sum() > 4000\n",
        "sim = sim.loc[~fil,~fil]\n",
        "sim"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0UCZhYrSWblV"
      },
      "source": [
        "Find the intersection between the genes in the similarity matrix and the genes in the embedding, reconstructed gene expression, or counts matrix.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "DG8EQbozVKaJ"
      },
      "outputs": [],
      "source": [
        "row_embedding_df = pd.DataFrame(embedded_genes, index=list(map(lambda x: x.upper(), node_features.index)))\n",
        "inter = row_embedding_df.index.intersection(sim.index)\n",
        "row_embedding_df = row_embedding_df.loc[inter]\n",
        "sim = sim.loc[inter,inter]\n",
        "row_embedding_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "S1HqxwPCVOqh"
      },
      "outputs": [],
      "source": [
        "out_features_df = pd.DataFrame(out_features.T, index=list(map(lambda x: x.upper(), node_features.index)))\n",
        "out_features_df = out_features_df.loc[inter]\n",
        "out_features_df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vQbh4TgmVSAx"
      },
      "outputs": [],
      "source": [
        "node_features_df = pd.DataFrame(node_features.values, index=list(map(lambda x: x.upper(), node_features.index)))\n",
        "node_features_df = node_features_df.loc[inter]\n",
        "node_features_df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e-MNryAjW3XS"
      },
      "source": [
        "Calculating the semantic similarity correlation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WuVdbU-sVVi4"
      },
      "outputs": [],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "from scipy.stats import rankdata, spearmanr\n",
        "\n",
        "def spearman_corr(matrix):\n",
        "    ranked_matrix = np.apply_along_axis(rankdata, 1, matrix)\n",
        "    return np.corrcoef(ranked_matrix)\n",
        "\n",
        "def corr_analysis(corr, sim):\n",
        "    sim_values = sim.values\n",
        "    correlations = np.empty(sim.shape[0])\n",
        "    for i in range(sim.shape[0]):\n",
        "        correlations[i], _ = spearmanr(corr[i], sim_values[i])\n",
        "    return np.abs(correlations)\n",
        "\n",
        "def plot_corr_distributions(corr_matrices, names, sim):\n",
        "    all_data = []\n",
        "\n",
        "    for matrix, name in zip(corr_matrices, names):\n",
        "        #corr = spearman_corr(matrix)\n",
        "        corr = matrix.values @ matrix.T.values\n",
        "        abs_corrs = corr_analysis(corr, sim)\n",
        "        all_data.extend([(name, val) for val in abs_corrs])\n",
        "\n",
        "    df = pd.DataFrame(all_data, columns=[\"Method\",\"Correlation\"])\n",
        "\n",
        "    custom_palette = {\n",
        "        names[0]: \"lightblue\",\n",
        "        names[1]: \"salmon\",\n",
        "        names[2]: \"lightgreen\"\n",
        "    }\n",
        "\n",
        "    plt.figure(figsize=(10, 6))\n",
        "    sns.violinplot(x=\"Method\", y=\"Correlation\", data=df, palette=custom_palette)\n",
        "    plt.title(\"Distribution of Spearman Correlations\")\n",
        "    plt.ylabel(\"Spearman r\")\n",
        "    plt.xlabel(\"Method\")\n",
        "    plt.xticks(rotation=45)\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "plot_corr_distributions([row_embedding_df,out_features_df, node_features_df], names=[\"Gene Embedding\",\"Reconstructed Gene Expression\", \"Counts\"], sim=sim)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DSYYzbl89CMN"
      },
      "source": [
        "## **Propagation based signature projection for actvation of Tcells**\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LmHplCk49QuF"
      },
      "outputs": [],
      "source": [
        "scNET.run_signature(recon_obj, up_sig=[\"Zap70\",\"Lck\",\"Fyn\",\"Cd3g\",\"Cd28\",\"Lat\"],alpha = 0.9)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "__JGbyqa9Sjn"
      },
      "source": [
        "**And for Tumor aggression**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EAFC9Ffox75N"
      },
      "outputs": [],
      "source": [
        "\n",
        "scNET.run_signature(recon_obj,up_sig=[\"Cdkn2a\",\"Myc\",\"Pten\",\"Kras\"])\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vamnJMrTxNdw"
      },
      "source": [
        "**Creating the co-embedded network, is it modular?**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Xs3IoUu5NiER"
      },
      "outputs": [],
      "source": [
        "import networkx as nx\n",
        "net, mod = scNET.build_co_embeded_network(embedded_genes, node_features)\n",
        "print(f\"The network mdularity: {mod}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "13MIgSdnC6uZ"
      },
      "source": [
        "##  Reconstract KEGG Pathways Example\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Aj-X3bEdDKu5"
      },
      "source": [
        "### Re-embed the T-cells subset"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8WiOIU8_ZuUJ"
      },
      "outputs": [],
      "source": [
        "obj = sc.read_h5ad(\"./example.h5ad\")\n",
        "sub_obj = obj[obj.obs[\"Cell Type\"] == \"CD8 Tcells\"]\n",
        "scNET.run_scNET(sub_obj, pre_processing_flag=False, human_flag=False, number_of_batches=3, split_cells=False, max_epoch=300, model_name = \"Tcells\")\n",
        "embedded_genes, embedded_cells, node_features , out_features =  scNET.load_embeddings(\"Tcells\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7KLaYR48MLMA"
      },
      "outputs": [],
      "source": [
        "from scNET import coEmbeddedNetwork  as cen\n",
        "import pandas as pd\n",
        "def run_pathway_recon_anaylsis(gene_embedding, ref, threshold=40):\n",
        "    '''\n",
        "    Predicts KEGG pathway memberships using gene embeddings and reference data, and evaluates the performance using AUPR.\n",
        "\n",
        "    Args:\n",
        "        gene_embedding (np.ndarray): The matrix of gene embeddings.\n",
        "        ref (pd.DataFrame): A reference dataset containing gene expression or other relevant features.\n",
        "\n",
        "    Returns:\n",
        "        pd.DataFrame: A DataFrame containing the AUPR scores for predictions from the gene embeddings and reference data.\n",
        "\n",
        "    Method:\n",
        "        - Annotates genes with KEGG pathway memberships using `crate_kegg_annot`.\n",
        "        - Filters KEGG pathways to include those with at least 40 gene members.\n",
        "        - Constructs co-embedded networks from both the embeddings and reference data.\n",
        "        - Uses propagation to predict pathway memberships for each graph.\n",
        "        - Evaluates the predictions using AUPR and plots the results.\n",
        "    '''\n",
        "    ref.index = list(map(lambda x: x.upper(),ref.index))\n",
        "    annot = cen.crate_kegg_annot(ref.index)\n",
        "    annot_threshold = annot.sum()>=threshold\n",
        "    annot_threshold = annot_threshold[annot_threshold == True].sort_values(ascending=False).head(50)\n",
        "    graph_embedded,_ = cen.build_co_embeded_network(gene_embedding,ref)\n",
        "    graph_ref,_ = cen.build_co_embeded_network(ref,ref)\n",
        "    kegg_pred = [cen.make_term_predication([graph_embedded,graph_ref], annot[term]) for term in annot_threshold.index]\n",
        "\n",
        "    kegg_pred = np.array(kegg_pred).squeeze()\n",
        "    df = pd.DataFrame({\"AUPR\" : kegg_pred.T.reshape(-1), \"Method\": [\"scNET\" for i in range(kegg_pred.shape[0])]  +  [\"Counts\" for i in range(kegg_pred.shape[0])]})\n",
        "\n",
        "    fig, ax = plt.subplots(figsize=[10,7])\n",
        "    fig.set_dpi(600)\n",
        "\n",
        "    custom_palette =  ['darkturquoise', 'lightsalmon']\n",
        "\n",
        "    sns.boxenplot(ax=ax, data=df,x=\"Method\", y=\"AUPR\", palette=custom_palette)\n",
        "    sns.set_theme(style='white',font_scale=1.5)\n",
        "    plt.show()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vKximdPpK_Fo"
      },
      "outputs": [],
      "source": [
        "\n",
        "run_pathway_recon_anaylsis(embedded_genes,node_features)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8pIsdt2oDZ5z"
      },
      "source": [
        "**Finding differential enriched pathways**\n",
        "### Can we see a difference in phenotype between Cancer, Microglia and Macrophages?\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Bs__k2sJbXHI"
      },
      "outputs": [],
      "source": [
        "recon_obj.obs[\"Cell Type\"] = recon_obj.obs.seurat_clusters.map(cell_types)\n",
        "de_genes_per_group, significant_pathways, filtered_kegg, enrichment_results = scNET.pathway_enricment(recon_obj.copy()[recon_obj.obs[\"Cell Type\"].isin([\"Microglia\",\"Macrophages\",\"Cancer\"])],groupby=\"Cell Type\")\n",
        "scNET.plot_de_pathways(significant_pathways,enrichment_results,10)"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "gpuType": "A100",
      "machine_shape": "hm",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "scNET",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.8"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}